### ISSUE TITLE
coverage starts too late with warning filter

### ISSUE THREAD

**Issue by dopplershift** [pytest-dev/pytest-cov#437]
# Summary

When trying to get coverage on something like `myproject`, while also running a warning filter on a warning within that project, say `myproject.MyWarning` coverage is started too late. For example:
```
python -m pytest -W error::myproject.MyWarning --cov=myproject tests/
```
will result in `myproject` being imported (to get `MyWarning`), before pytest-cov starts coverage. Any imports (and code that runs at import time) as a results of importing `myproject` will be missed. You do get the warning: `Coverage.py warning: Module myproject was previously imported, but not measured (module-not-measured)`

I'm not sure if anything can be done about it (since pytest is the one controlling when warning filters are set up), but at the very least it would be great to put this as a gotcha in the docs somewhere.

**Comment by ionelmc** [comment]
I guess this section should be extended to outline your usecase as well (it's the same solution): https://pytest-cov.readthedocs.io/en/latest/plugins.html

**Comment by dopplershift** [comment]
Yeah, that might have helped. I definitely didn't think to look at that section.

**Comment by flying-sheep** [comment]
I wonder if there is a better solution possible.

- if pytest is interested in fixing this, we can use this test: https://github.com/flying-sheep/pytest/commit/f53823569008ca6490ce9da7036a93e0d129f19c

  An approach to fix it in pytest would need to avoid importing the package under coverage during plugin initialization or earlier. So we could add a `delayed_imports = ['pkg1']` config or something like that which causes `filterwarnings` entries mentioning these packages to only be applied in later stages (e.g. in the [`pytest_runtest_protocol` hook](https://github.com/pytest-dev/pytest/blob/089116bdff4a82b624caa50913910e11cfad2d51/src/_pytest/warnings.py#L106-L111), but not in the [`pytest_load_initial_conftests` hook](https://github.com/pytest-dev/pytest/blob/089116bdff4a82b624caa50913910e11cfad2d51/src/_pytest/warnings.py#L143-L150))

- another solution would be to add a `pytest_cov.warnings` plugin or so that can be used instead of `warnings` (`addopts = ['no:warnings', 'pytest_cov.warnings']`) and given `--cov=myproj` will remove `myproj.*` filters.

- a combined solution would be for pytest to allow other plugins to hook into the `filterwarnings` processing of the `warnings` plugin, so `pytest-cov` can patch in that behavior.
